### ISSUE TITLE
QoL: Functionality to allow more than one dictionary object literal declarations of the same type and name (such as theme dictionaries) in JSON Intrepreter/Compiler

### ISSUE THREAD

**Issue by samusxx** [microsoft/vscode#142370]
This is a _QoL issue_ to be _handled_ by the JSON Interpreter/Compiler (both for realtime interpretation in the VSCode editor and compiled in any other situation).

### Functionality to allow more than one (i.e. 2+) declaration in a JSON file, _of the same dictionary name and type_.
(The example case is themes like `textMateRules` in `editor.tokenColorCustomizations` )
The goal is to have the JSON Interpreter and Compiler understand the written json as suggested is _simply an extension of the first declaration_. The Interpreter (if that's what it's called) for realtime changes in the editor, consistent with everything else, and also the Compiler to make it consistent in all cases.

### Why? 

Because some `"workbench.colorCustomizations" : {}` content overrides* some colors in `"editor.tokenColorCustomizations" : { textMateRules : {} }`, and you can't declare  `editor.tokenColorCustomizations` or `textMateRules` afterward, if they're already defined. If undefined, doing so just reverses the situation between the two theme, object literal dictionary declarations anyway.

This would allow the ability to do the `textMateRules` theming, then set the necessary `workbench.colorCustomizations`, and if there are conflicts or overrides of `textMateRules` theming, then you can simply redeclare it and the JSON Interpreter and Compiler will understand it's simply an extension of the first declaration.

**Alternatively:** _Self-referencing_ the object literal declarations/initializers _within the same JSON file_, would be an acceptable resolution for this kind of problem.

*_If this shouldn't happen, I've filed a bug report about `"editorBracketHighlight.foreground1"-3` in `workbench.colorCustomizations`, because it _does_ override `foreGround` colors on `"scope" : "meta.brace.square"` in `"editor.tokenColorCustomizations" : { "textMateRules" : {} }`._ 

Steps to Reproduce:

1. For themes, in settings.json, create `"editor.tokenColorCustomizations" : { "textMateRules" : {} }"` object literal dictionary structure declaration and initialization as normal. Create a scope for `meta.brace.square` and set its `foreGround` color.
2. Similarly, then create `"workbench.colorCustomizations" : {}"` below it. In this structure (2.), set colors for `"editorBracketHighlight.foreground"1-3`.
4. Add `"editor.bracketPairColorization.enabled": true`, on one line of the settings.json. Save.
5. Create a new javascript file and create several, nested brace pairs of a normal type, then several square brace pairs inside the innermost normal braces.
6. Observe that the square braces are changed by Step 3. and Step 2. You still want the normal braces `( ( () ) )` to be colored by nesting.
7. Now you can't get the _square_ braces to be colored while `"editor.bracketPairColorization.enabled": true`. Placing Step 1. after Step 2. doesn't work either even though the _Developer: Inspect Editor Tokens and Scopes **tooltip**_ recognises the `foreGround` color being set on the **square braces**. Even if it worked this way, it reverses the problem mentioned earlier anyway, whereby it _may_ occur that some `workbench.colorCustomizations` are replaced by `textMateRules` that are declared afterward (untested, but possible) and you would want to re-set because of load order, these replaced `workbench.colorCustomizations`.

**Comment by samusxx** [comment]
Had to edit due to a mistake. (sorry). The goal was to have nested coloring on normal braces `( ( () ) )`, but then ensure that you can color square braces `[ ]` separately, which isn't respected even in the case of switching the two theme, object literal dictionary declarations orders, which still retains the one overriding the former in unintended ways as the two theme object literals are contiguous structures, you can't pick out one 'key'/element and put it later.

**Comment by samusxx** [comment]
Another ALTERNATIVE is to have a **priority** property on _`textMateRules`_ for each _`scope`_, whereby **`"priority" = 1`** (or any other integer for another position) sends it to the _last position to be executed_ in realtime for editor purposes and compiled to trigger last. The debugger could detect priority=1 applying to the same _scope_ and suggest a 2 instead or, for the user to comment out the undesired conflict.

**Comment by samusxx** [comment]
Idea requires the JSON standard to be updated.

**Comment by aeschli** [comment]
The bracket pair coloration feature is a feature on top of the syntax highlighting. 

All colors for the bracket pair coloration feature should all be fully definable by the `workbench.colorCustomizations` setting and they go over the colors defined by `textMateRules`.

The order in the JSON file of `workbench.colorCustomizations` and `editor.tokenColorCustomizations` should not matter.

Can't you set the colors you want all in `workbench.colorCustomizations` ?

**Comment by hediet** [comment]
> Functionality to allow more than one (i.e. 2+) declaration in a JSON file, of the same dictionary name and type.

This is out of scope for this project.

**Comment by samusxx** [comment]
Look, how about as a little fix, just copy+paste the code for bracketpaircolorizations and rename one of them square braces, and also make it only apply to square braces, while making the other copy only apply to curly braces? When the 3 colors are defined in workbench, then you can apply them only if 'SquarebracketColorisations is true'. That sounds easy huh?

There really is no point continuing bracketpaircolorizations for square braces. They are never far apart. 

This won't change any other problems with workbench overriding.

**Comment by samusxx** [comment]
@hediet 
a. yes.
b. yes.

**this is better:**
c. The square braces `[ ]` colorization, in the `workbench.colorCustomizations` settings, should be separated out of `"editor.bracketPairColorization.enabled"`.

_The reasoning for this is, you don't need nested pair colorization on square braces_, but you sitill want it on regular...why? _because square braces are never far apart_, or rarely. I want to see a different color on them while `"editor.bracketPairColorization.enabled"`  is true and so would many others.

**Implementation can be easier than you think.** 
1. Simply remove the square braces from `"editor.bracketPairColorization.enabled"`. This immediately doesn't overwrite the `"scope" : "meta.braces.square"`.  _Resolved_
_Optionally also:_
2. If you want to allow nested colorization of square braces still, make a duplication of the code `"editor.bracketPairColorization.enabled"` to be `"editor.**square**BracketPairColorization.enabled"`, also still using `"editorBracketHighlight.foreground1"-3`. The latter can still use the 3 settings for normal braces it just is now selectively implemented between both `( )` and `[ ]` , so nothing more needs to change.  _Resolved_
_Optionally also:_
 3. Make `"editor**Square**BracketHighlight.foreground1"-3` to apply to nested `[ ]` square braces. Then `"editorBracketHighlight.foreground1"-3` needs to be set only to normal braces `( )`.  _Resolved_
  

_**OR**_
A. Make a new property of `"TextMateRules" : {  { "settings" } }` called `priority`. This would be used for the final color determination above workbench, if `1`. And could also be useful for any other situation, because it would place the execution of the piece last (either by compiler, interpreter, or at a higher code level).  _Resolved_

_**OR**_
B. Allow multi-declarion of JSON object literals, where they extend the first declaration, rather than override it. _Resolved_
Known issue: _This may be problematic for any situation for overwriting the first declaration etc. Not sure how to resolve this complication._  

So choose, 1.
Or choose, 1+2.
Or choose, 1+2+3.
Or A or B, alternatively.

**Comment by samusxx** [comment]
By the way the idea of duplicate properties is becoming more accepted. There's no reason for them to do this recently, but they did. Whereas my suggestion has a purpose. 
![Ow1JH8t8cb](https://user-images.githubusercontent.com/32408456/153561153-91485e9f-88db-42b1-9456-83273e00be15.png)
